[AC-2579] Set up bit-cli folder

eliykat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

<!--Describe what the purpose of this PR is. For example: what bug you're fixing or what new feature you're adding-->

We're now adding Bitwarden Licensed code to the CLI client, so we need a separate `bit-cli` folder and build.

This PR does 2 things:

1. creates the `bitwarden_license/bit-cli` folder, config files, and entrypoint, and adds `bit` commands to `apps/cli/package.json`. This largely follows the `bit-web` structure.
2. refactors the CLI app to give more flexibility in registering arbitrary Programs depending on the entry point

Setting up CI builds is out of scope and will be done separately.

## Code changes

<!--Explain the changes you've made to each file or major component. This should help the reviewer understand your changes-->
<!--Also refer to any related changes or PRs in other repositories-->

**Tools team:** the only changes to your files are updating imports and naming.

**Architecture devs:** I would like you to review the substantive changes here.

The `bit-cli` config files and the new `package.json` commands are fairly self-explanatory, although they still need to be checked over.

The changes to the CLI app structure require a bit more explanation. At the moment, our services (from `libs/common`) and the CLI program classes are tightly coupled in the `Main` class, which is responsible for instantiating services, registering programs, and starting execution of the app. This is too many responsibilities and is quite rigid.

The refactored code takes advantage of Commander's builder pattern to give us more flexibility. The new code works like this:
1. `ServiceContainer` is responsible for initializing all services and making them available to programs (usually passed into their constructors)
2. `registerOssPrograms` is responsible for initializing and registering all (OSS) programs
3. `bw.ts` stitches it together: instantiate the service container, pass it to `registerOssPrograms`, start the app.

bit-cli has an identical structure:
1. `ServiceContainer` subclasses the OSS class and allows for registration of additional bit-licensed services
2. `registerBitPrograms` registers bit-licensed programs
3. `bw.ts` stitches it together, note that it calls `registerOssPrograms` first so that bit-cli is always a superset of the OSS functionality.

## Screenshots

<!--Required for any UI changes. Delete if not applicable-->

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

eliykat commented on apps/cli/src/service-container.ts:
This file is actually just `bw.ts` renamed, with the `run()` method and all program classes removed. There are no changes to the rest of the file, e.g. how services are bootstrapped.

eliykat commented on bitwarden_license/bit-cli/src/bw.ts:
```diff
@@ -0,0 +1,20 @@
+import { registerBitPrograms } from "./register-bit-programs";
+import { ServiceContainer } from "./service-container";
+
+async function main() {
```
This (and its OSS counterpart) does the work of what `Main.run()` used to do:
* instantiate services
* instantiate and register programs
* kick off execution.

One change is that we used to explicitly check for no arguments and show help text: https://github.com/bitwarden/clients/blob/main/apps/cli/src/bw.ts#L740-L742. However, Commander does this automatically since version 5 and we're now on version 11: [SO reference](https://stackoverflow.com/a/70458436), [Github reference](https://github.com/tj/commander.js/issues/1088#issuecomment-580988629). I've removed it, which also has the nice effect of removing _any_ actual logic from `main`, it's all just initialization.

eliykat commented:
I've refactored this since the draft yesterday and updated the description above.

audrey-jensen reviewed (dismissed):
Tools changes LGTM.

Kudos 👍🏻 - Loving how simple the entry point is now.
Thoughts 💭 - I'm guessing license issues don't apply to us due to CLAs, but it's still a little curious that we directly derive from the CLI's service class.
Nitpick ⛏️ - Can any of this be unit tested?

MGibson1 commented on apps/cli/package.json:
Please update normal build scripts to `build:oss` (see web scripts as pattern). If this is intended as part of your automation follow up, that's fine, too.

eliykat replied:
Will do! I avoided changing any existing scripts because that will require updates to Github workflows, which I didn't want to deal with here. But I will include it in the follow-up.

MGibson1 reviewed (commented):
This looks just about perfect! well done. I want to understand the plan on script updates, but depending on your answer I'll approve as is.

eliykat commented:
Additional changes:

1. resolved merge conflicts - I had to manually reapply some changes to `ServiceContainer`, these conflicting PRs are referenced in the commits above
2. added `bit (cli)` to the vscode workspaces for those that use this feature
3. made `inited` a private property (something I added but forgot to mention - now that `ServiceContainer.init()` is public we want to make sure it's not called more than once)
4. added an _extremely_ minimal unit test around `ServiceContainer`, at least to make sure it instantiates properly - although given that we have to manually order and instantiate our services in its constructor, this is not actually a given. It would be nice to test `init()`, but several services interact directly with the filesystem via `fs`, which would require more time setting up mocks than I want to give it here.

audrey-jensen commented on apps/cli/src/service-container.ts:
```diff
@@ -0,0 +1,770 @@
+    process.env.BW_SESSION = null;
+  }
+
+  async init() {
+    if (this.inited) {
+      return;
```
🎨 This should probably log when it's called more than once; that way we'll get alerted when this invariant fails.

eliykat replied:
I wasn't sure what level this should be - I went with `LogLevel.Warning`, but I can fix it in a subsequent PR if required.

audrey-jensen replied:
It's not broken, but it is malfunctioning. I think `Warning` communicates that. 😄

audrey-jensen reviewed (dismissed):
🎨 suggestion
